src/Engine/ProtoCore/DSASM/Heap.cs

```diff
@@ -416,7 +424,28 @@ public StackValue AllocatePointer(int size, MetaData metadata)
         private StackValue AllocateStringInternal(string str, bool isConstant)
         {
             int index;
-            if (!stringTable.TryGetPointer(str, out index))
+            if (stringTable.TryGetPointer(str, out index))
+            {
+                // Any existing heap elements, marked as white, that are in the sweepSet and that are referenced during the sweep cycle will be marked as Black.
+                // This will ensure that no reachable data is mistakenly cleaned up.
+                bool isDuringGCCriticalAsyncCycle = gcState != GCState.Pause;// Between the time the GC takes a snapshot of the stack and heap untill GC cycle is over.
+                bool isValidHeapIndex = index >= 0 && index < heapElements.Count;
+                if (isDuringGCCriticalAsyncCycle && isValidHeapIndex)
+                {
+                    var he = heapElements[index];
+                    Validity.Assert(he != null, $"Null heap element found at index {index} during AllocateStringInternal");
+
+                    // If heap element is marked as white then it is either not processed by Propagate step yet or processed and found as garbage.
+                    // If the sweepSet does not contain the heap element's index then there is no need to mark it black (since cleanup will not even be tried)
+                    if (he.Mark == GCMark.White && sweepSet.Contains(index))
```

Comment:
Does this imply that a heap element will be cleaned up if both are true, i.e., it is marked white and it is contained in the sweep set? If it is contained in the sweep set, does it mean that it is processed and found as garbage?

Reply:
`Does this imply that a heap element will be cleaned up only if both are true, i.e., it is marked white and it is contained in the sweep set?`
Yes, both need to be true for a heap element to be cleaned up by the GC.

`If it is contained in the sweep set, does it mean that it is processed and found as garbage?`
No. The sweepSet represents the heap elements that are to be considered for cleaning. Not all heap elements are allowed to be cleaned up (ex. constant strings - **fixedHeapElements** and heap elements that have already been cleaned up - **freeList** - and not yet re-used)

there are 2 ways to **not clean up** a heap element. Keep it out of the SweepSet or mark it as gray/black during GC cycle.  Both mechanism are used because of different lifecycles (colors will be reset after each gc cycle....but **freeList** and **fixedHeapElements** persist)